### ISSUE TITLE
Making formlets serializable

### ISSUE THREAD

**Issue by LiberalArtist** [racket/web-server#23]
I wanted to revisit the question raised in [this old mailing-list thread](https://groups.google.com/d/msg/racket-users/lMYWjodgpmo/ib7KOa80nRAJ) about using `#lang web-server/base` for the formlets library so that formlets would be serializable.

The primary reason I personally would want this is because I often want to generate formlets dynamically, e.g. using something like `checkbox-group` to have a user select one of the results of a database query.

Is there a significant downside to doing this?

**Comment by jeapostrophe** [comment]
There are some downsides.

First, the `#lang web-server` compiler is slow and the programs that come out are slower.

Second, if you apply contracts to something from `#lang web-server` then the continuation is not serializable, which defeats the purpose of doing this, but then that means that we have to remove the contracts on formlets, which seems painful.

In situations like this, I really want to be able to do something like `(require (with-new-language web-server/formlets web-server/base))` and force the other module to work differently. There's all sorts of obvious problems with this, of course.

At a first pass, does `web-server/formlets` compile under `#lang web-server/base`? Can you test it?

**Comment by LiberalArtist** [comment]
It compiles and works with `#lang web-server/base`. The only miscellaneous change is that instead of using `const` from `racket/function`, `web-server/formlets/lib` needs to e.g.
```
(define (const x)
  (λ _ x))
```

Contracts seem like the main sticking point to me.

I don't know if this distinction helps in this case, but IIUC there are two seperate possible benefits to having formlets be in `#lang web-server/base`: being able to call `send/suspend` and friends within a formlet-processing procedure would be one, and that would need the relevant continuations to be serializable. But it would also be useful just to have formlets be serializable as values within the continuation (and that is the part that's more essential to work with dynamically-generated formlets). I suspect that may still be a problem for contracts, though, because I suspect that a serializable procedure with a contract would deserialize to the contractless version. My understanding of impersonators and chaperones is not especially deep, though.

**Comment by LiberalArtist** [comment]
Here's what it looks like with `#lang web-server/base` but no contracts: https://github.com/LiberalArtist/web-server/commit/93093d2349877f83c7ce8d758a0c7948c3514879

**Comment by LiberalArtist** [comment]
As an experiment, I tried making wrapper procs to do checking on formlets that are part of the domain of the functions in `web-server/formlets/lib`. It's tedious and definitely worse than real contracts, but I think it could work for the rest of the library, too. https://github.com/LiberalArtist/web-server/commit/c05d5e87c62d88080d613a5521b167d5765dfefc

**Comment by LiberalArtist** [comment]
After more experimentation, I became less happy with the manual wrappers I discussed above: aside from being painful and error-prone  (given that they were re-implementing lots of checks that the contract system does very well), there wasn't a good story for what programmers outside of the formlets library should be doing to address these issues.

In https://github.com/racket/web-server/pull/30, I decided having formlets be a serializable data structure (so dynamically-generated formlets can be serialized as part of a closure) was worth addressing as a seperate issue from being to capture continuations during the processing stage of a formlet. It turned out to be fairly easy to just write the formlets implementation in `#lang web-server/base` (with a few other minor changes) and have all of the combinators in the existing library produce formlets that can be serialized (as long as they only contain serializable pieces, as usual for serialization).

I also created a `web-server/formlets/unsafe` that has no contract checking, thus allowing continuation capture, but in practice I've found that just being able to construct serializable formlets dynamically is enough to deal with the cases that led me to start this discussion.
